### ISSUE TITLE
The solution of a quest hides an unrelated (nearby) quest

### ISSUE THREAD

**Issue by deevroman** [streetcomplete/StreetComplete#4371]
**How to Reproduce**

https://user-images.githubusercontent.com/25708359/189550770-c4ca51e6-db81-4ae8-84ad-4c6c5e66d93e.mp4

**Expected Behavior**

**Versions affected**

47.0 alpha1

Samsung A52 Android 12. Samsung J1 Android 5.1.1

**Comment by dbdean** [comment]
I've experienced this with any quest on a way that also has quests on nodes. You need to rescan for quests to get them to re-appear. (Sorry for not reporting this earlier!)

**Comment by cyclingcat** [comment]
> I've experienced this with any quest on a way that also has quests on nodes. You need to rescan for quests to get them to re-appear. (Sorry for not reporting this earlier!)

The same here, for many months now. This bug definitely hasn't been introduced with the overlays feature, it has been present for a much longer time. I have tried to reproduce it several times but haven't succeeded so far - it happens often, but not always and I haven't found a pattern behind it yet.

The cycling cat

**Comment by westnordost** [comment]
Well, I can't reproduce this, but it would help to see the log when this happens.

Alternatively, try exiting the app (tab out, remove from recents) and restart. Is the quest still gone?

When you solve a quest on a way, actually what **should** happen is that only for that way, it is re-checked which should be displayed (now). That a quest referring to a different element vanishes, is peculiar.

**Comment by Helium314** [comment]
I think the missing quest is not necessarily connected to quests on ways. It's only noticeable there because when the road changes is a re-scan for quests happening.
I also had other quests missing, though that is very hard to notice. E.g once I found a shop without name in nearby highlighted elements but no "name" quest. On re-scan the "name" quest suddenly showed up... and this was _not_ near the edge of the downloaded area, where such things may happen.

**Comment by deevroman** [comment]
Please note that the other quest disappears only after sending the changes.


https://user-images.githubusercontent.com/25708359/190828849-811fb803-f6ae-425f-a22f-840d676ed941.mp4

[logcat_09-17-2022_01-56-18.txt](https://github.com/streetcomplete/StreetComplete/files/9590465/logcat_09-17-2022_01-56-18.txt)


p.s. SC is lagging due to recording on an old device (Samsung J1, Android 5.1.1) But the bug is also playing on my main Samsung A52 Android 12 device

**Comment by westnordost** [comment]
Some good information right there, i.e. that it only occurs after upload. The log however is not helpful, there should be logs by the `QuestController`.

**Comment by deevroman** [comment]
how can I get them?

**Comment by westnordost** [comment]
They should be at the same place. Maybe you cut if off too early?

**Comment by westnordost** [comment]
Actually, there are no StreetComplete logs at all in that file. Maybe LogCat  Reader only captures the system logs or something? I don't know about that app.

**Comment by deevroman** [comment]
okay, I built an application from this repository and this is what came out in the console

[AndroidStudio.log](https://github.com/streetcomplete/StreetComplete/files/9592528/AndroidStudio.log)

**Comment by westnordost** [comment]
```
I/Upload: Starting upload
D/OsmQuestController: AddMaxHeight requires surrounding map data to determine applicability to WAY#1095254741
I/MapDataController: Fetched 39 elements and geometries in 161ms
I/OsmQuestController: Created 0 quests for 1 updated elements in 0.7s
I/OsmQuestController: Persisted 0 new and removed 0 already resolved quests in 0.0s
I/ChangesetManager: Created changeset #126311433
D/ElementEditsUploader: Uploaded a UpdateElementTagsAction
D/OsmQuestController: AddMaxHeight requires surrounding map data to determine applicability to WAY#1095254741
I/MapDataController: Fetched 39 elements and geometries in 45ms
I/OsmQuestController: Created 0 quests for 5 updated elements in 0.1s
I/OsmQuestController: Persisted 0 new and removed 3 already resolved quests in 0.0s
I/Upload: Finished upload
```

**Comment by westnordost** [comment]
Hmm, two things are suspect.

1. why does the `OsmQuestController` create and persist quests apparently **before** the `ElementEditsUploader` finished uploading (even before creating the changeset)
2. Why are the 4 updated elements? Apparently only of one element, the tags were changed. Why are there now 5 elements that have been updated?

**Comment by deevroman** [comment]
Maybe it's because I canceled edits on this site.
---
Is it possible to switch StreetComplete to the osm sandbox? It is not very convenient to test on real objects

**Comment by westnordost** [comment]
> Is it possible to switch StreetComplete to the osm sandbox? It is not very convenient to test on real objects

Only if you compile it yourself. Change `OSM_API_URL` in OsmApiModule

**Comment by westnordost** [comment]
I found:

> 1. why does the OsmQuestController create and persist quests apparently before the ElementEditsUploader finished uploading (even before creating the changeset)

This is actually normal, the quest solution is applied locally. When auto-upload is on, it will be uploaded immediately afterwards.

> 2. Why are the 4 updated elements? Apparently only of one element, the tags were changed. Why are there now 5 elements that have been updated?

I also found the reason why this is happening. Will provide a fix...

**Comment by westnordost** [comment]
**The explanation as provided in the commit message:**

> on upload, only broadcast the actually updated element (fixes #4371)

>when the OSM API returns the updated element, the uploader fetches other elements from the local db that are necessary to construct the (new) geometry of the updated element (i.e. nodes of a way, member ways+nodes of a relation). Now, these elements have not actually been updated, they are just used to create the geometry. Hence, only the elements returned from the API must be broadcast as having been updated.
> 
> The issue identified in #4371 came to pass because that hasn't been the case but at the same time, no geometry for those additional elements was provided and so all the quests referring to elements with apparently missing geometry were removed.

This must have been a long-standing bug, well over a year old. Thank you for reporting it and providing the hints that made me find it!
